pkg/pma/pma.c

```diff
@@ -0,0 +1,860 @@
+static int
+handle_page_fault_(void *fault_addr, void *user_arg);
+
+/// Handle SIGSEGV according to the libsigsegv protocol. See sigsegv_handler_t
+/// in sigsegv.h for more info.
+///
+/// @param[in] fault_addr
+/// @param[in] serious
+static int
+handle_sigsegv_(void *fault_addr, int serious);
+
+/// Map a file into memory at a specific address. Can also be used to create an
+/// anonymous, rather than file-backed, mapping by passing in a NULL path.
+///
+/// @param[in]     path        Path to the backing file. If NULL, an anonymous
+///                            mapping is created.
+/// @param[in]     base        Base address of the new mapping.
+/// @param[in]     grows_down  true if the mapping should grow downward in
+///                            memory.
+/// @param[in]     pma         PMA this mapping will belong to.
+/// @param[in,out] len         Populated with the length of the new mapping. If
+///                            path is NULL, this parameter can also be used to
+///                            supply a non-default length for an anonymous
+///                            mapping.
+/// @param[out]    fd          Populated with the file descriptor of the opened
+///                            backing file. -1 if path is NULL.
+///
+/// @return 0   Successfully created a new mapping.
+/// @return -1  Failed to create a new mapping.
+static int
+map_file_(const char *path,
+          void       *base,
+          bool        grows_down,
+          pma_t      *pma,
+          size_t     *len,
+          int        *fd);
+
+/// Get the status of the page surrounding an address. To set rather than get,
+/// see set_page_status_().
+///
+/// @param[in] addr  Address within the page in question. Must be within the
+///                  bounds of the PMA.
+/// @param[in] pma   PMA the page in question belongs to.
+static inline page_status_t
+page_status_(const void *addr, const pma_t *pma)
+{
+    size_t  pg_idx    = addr_to_page_idx_(addr, pma);
+    size_t  entry_idx = pg_idx / kPagesPerEntry;
+    size_t  bit_idx   = (pg_idx % kPagesPerEntry) * kBitsPerPage;
+    uint8_t status    = (pma->pg_status[entry_idx] >> bit_idx) & PS_MASK;
+    return status;
+}
+
+/// Set the status of the page surrounding an address. To get rather than set,
+/// see page_status_().
+///
+/// @param[in] addr    Address within the page in question. Must be within the
+///                    bounds of the PMA.
+/// @param[in] status  New status of the page in question.
+/// @param[in] pma     PMA the page in question belongs to.
+static inline void
+set_page_status_(const void *addr, page_status_t status, const pma_t *pma)
+{
+    size_t  pg_idx    = addr_to_page_idx_(addr, pma);
+    size_t  entry_idx = pg_idx / kPagesPerEntry;
+    size_t  bit_idx   = (pg_idx % kPagesPerEntry) * kBitsPerPage;
+    uint8_t entry     = pma->pg_status[entry_idx];
+    pma->pg_status[entry_idx]
+        = (entry & ~(PS_MASK << bit_idx)) | (status << bit_idx);
+}
```

Comment:
can't this just be accomplished with some bit arithmetic? -- trying to avoid idiv since this is a hotpath -- not nearly as hot as the main bytecode interpreter loop, but still ... not cold. IDK if there will be any real world observable difference, but I also think it aids readability. At least, I understand the page status tracking logic better when we partition the page address bitwise. 

The pg_idx is just the number of bits beyond number of bits in a page (12 for 4K page). 
Entry idx should just be pg_idx >> 2. 
bit_idx you can just mask pg_idx with 3 and mul by kBitsPerPage.

Reply:
The division and mod operations are replaced by `shr` and `and`, respectively when compiling with `-O2` on `linux-x86_64`, so there's no effective functional difference when compared to bit shifting. We could spend a lot of time discussing which approach more clearly represents the logic, but because the resulting functionality is identical in both cases, I don't think that's a good use of our time.